File: backend/src/services/legalOntology.service.ts
// =====================================================
// 🔥 SOLVELITIGATION LEGAL ONTOLOGY ENGINE (UPGRADED)
// =====================================================

export type LegalCategory =
  | "Criminal"
  | "Civil"
  | "Service Law"
  | "Taxation & Corporate"
  | "Unknown";

// =====================================================
// 🔥 CATEGORY NORMALIZATION
// =====================================================

const CATEGORY_NORMALIZATION:
Record<string, LegalCategory> = {

  "criminal": "Criminal",
  "criminal law": "Criminal",
  "crime": "Criminal",

  "civil": "Civil",
  "civil law": "Civil",
  "constitutional": "Civil",

  "service": "Service Law",
  "service law": "Service Law",
  "servicelaw": "Service Law",

  "taxation": "Taxation & Corporate",
  "corporate": "Taxation & Corporate",
  "tax": "Taxation & Corporate",

  "unknown": "Unknown"
};

// =====================================================
// 🔥 NORMALIZE CATEGORY
// =====================================================

export const normalizeCategory = (
  rawCategory: string = ""
): LegalCategory => {

  const cleaned =
    rawCategory
      .trim()
      .toLowerCase();

  return (
    CATEGORY_NORMALIZATION[cleaned]
    || "Unknown"
  );
};

// =====================================================
// 🔥 CATEGORY KEYWORDS
// =====================================================

const CATEGORY_KEYWORDS:
Record<LegalCategory, string[]> = {

  "Criminal": [

    "ipc",
    "bns",
    "bnss",
    "crpc",
    "criminal",
    "crime",
    "murder",
    "rape",
    "dowry",
    "ndps",
    "pocso",
    "bail",
    "fir",
    "chargesheet",
    "investigation",
    "conviction",
    "acquittal",
    "sentence",
    "evidence",
    "cheque bounce",
    "negotiable instruments",
    "pmla",
    "money laundering",
    "prevention of corruption",
    "uapa",
    "terror",
    "narcotic"
  ],

  "Civil": [

    "civil",
    "contract",
    "property",
    "partition",
    "succession",
    "specific relief",
    "transfer of property",
    "injunction",
    "lease",
    "rent",
    "arbitration",
    "family",
    "marriage",
    "consumer",
    "motor vehicle",
    "wakf",
    "title dispute",
    "civil procedure",
    "constitution",
    "article 14",
    "article 21"
  ],

  "Service Law": [

    "service",
    "departmental",
    "disciplinary",
    "promotion",
    "seniority",
    "pension",
    "reinstatement",
    "termination",
    "dismissal",
    "army",
    "navy",
    "air force",
    "civil services",
    "administrative tribunal",
    "departmental inquiry",
    "charge memo",
    "service benefits",
    "conduct rules",
    "ccs",
    "cat"
  ],

  "Taxation & Corporate": [

    "gst",
    "income tax",
    "customs",
    "excise",
    "tax",
    "assessment",
    "reassessment",
    "refund",
    "input tax credit",
    "corporate",
    "companies act",
    "ibc",
    "insolvency",
    "nclt",
    "liquidation",
    "shareholder",
    "board resolution",
    "sebi",
    "fema",
    "rbi",
    "sarfaesi",
    "banking"
  ],

  "Unknown": []
};

// =====================================================
// 🔥 INVALID CROSS-ONTOLOGY TOPICS
// =====================================================

const INVALID_COMBINATIONS = [

  ["murder", "arbitration"],
  ["ndps", "tenancy"],
  ["insolvency", "bail"],
  ["partition", "cheque bounce"],
  ["service", "murder"],
  ["corporate", "rape"]
];

// =====================================================
// 🔥 NORMALIZE ACT NAME
// =====================================================

export const normalizeActName = (
  actName: string = ""
): string => {

  return actName
    .replace(/\s+/g, " ")
    .replace(/[.,]/g, "")
    .trim()
    .toLowerCase();
};

// =====================================================
// 🔥 CASE NUMBER VALIDATOR
// =====================================================

export const isValidCaseNumber = (
  caseNumber: string = ""
): boolean => {

  const text =
    caseNumber.toLowerCase();

  if (
    text.includes("unknown case") ||
    text.includes("item no") ||
    text.includes("notice")
  ) {
    return false;
  }

  return text.length >= 10;
};

// =====================================================
// 🔥 INFER CATEGORY FROM ACT
// =====================================================

export const inferCategoryFromAct = (
  actName: string = ""
): LegalCategory => {

  const normalized =
    normalizeActName(actName);

  let bestCategory:
  LegalCategory = "Unknown";

  let bestScore = 0;

  for (
    const [category, keywords]
    of Object.entries(CATEGORY_KEYWORDS)
  ) {

    let score = 0;

    for (const keyword of keywords) {

      if (
        normalized.includes(
          keyword.toLowerCase()
        )
      ) {
        score++;
      }
    }

    if (score > bestScore) {

      bestScore = score;

      bestCategory =
        category as LegalCategory;
    }
  }

  return bestCategory;
};

// =====================================================
// 🔥 CASE NUMBER CATEGORY
// =====================================================

export const inferCategoryFromCaseNumber = (
  caseNumber: string = ""
): LegalCategory => {

  const text =
    caseNumber.toLowerCase();

  if (
    text.includes("criminal") ||
    text.includes("crl") ||
    text.includes("bail")
  ) {
    return "Criminal";
  }

  if (
    text.includes("service")
  ) {
    return "Service Law";
  }

  if (
    text.includes("tax") ||
    text.includes("gst")
  ) {
    return "Taxation & Corporate";
  }

  if (
    text.includes("civil") ||
    text.includes("writ")
  ) {
    return "Civil";
  }

  return "Unknown";
};

// =====================================================
// 🔥 SECTION FALLBACK
// =====================================================

export const inferCategoryFromSections = (
  sections: any[] = []
): LegalCategory => {

  for (const s of sections) {

    const act =
      String(
        s?.act || ""
      );

    const inferred =
      inferCategoryFromAct(act);

    if (inferred !== "Unknown") {

      return inferred;
    }
  }

  return "Unknown";
};

// =====================================================
// 🔥 FILTER INVALID TOPICS
// =====================================================

export const removeInvalidOntologyTopics = (
  topics: string[] = []
): string[] => {

  const cleaned =
    topics.map(t =>
      t.toLowerCase()
    );

  for (const pair of INVALID_COMBINATIONS) {

    const hasA =
      cleaned.some(t =>
        t.includes(pair[0])
      );

    const hasB =
      cleaned.some(t =>
        t.includes(pair[1])
      );

    if (hasA && hasB) {

      return topics.filter(
        t =>
          !t.toLowerCase().includes(pair[1])
      );
    }
  }

  return topics;
};

// =====================================================
// 🔥 FILTER ACTS BY CATEGORY
// =====================================================

export const filterActsByCategory = (
  acts: string[] = [],
  category: string = ""
): string[] => {

  return acts.filter(act => {

    const inferred =
      inferCategoryFromAct(act);

    return inferred === category;
  });
};

// =====================================================
// 🔥 DOCUMENT QUALITY ENGINE
// =====================================================

export const calculateDocumentQuality = (
  judgment: any
): string => {

  const caseNumber =
    String(
      judgment?.caseNumber || ""
    );

  const headnote =
    String(
      judgment?.headnote || ""
    );

  const acts =
    judgment?.acts || [];

  if (
    !isValidCaseNumber(caseNumber)
  ) {
    return "LOW_QUALITY";
  }

  if (
    headnote.includes("para-unknown")
  ) {
    return "LOW_QUALITY";
  }

  if (
    acts.includes("Unknown Act")
  ) {
    return "LOW_QUALITY";
  }

  return "HIGH_QUALITY";
};

// =====================================================
// 🔥 REVIEW REQUIRED
// =====================================================

export const requiresManualReview = (
  judgment: any
): boolean => {

  if (
    judgment?.category === "Unknown"
  ) {
    return true;
  }

  if (
    judgment?.documentQuality ===
    "LOW_QUALITY"
  ) {
    return true;
  }

  if (
    judgment?.acts?.includes(
      "Unknown Act"
    )
  ) {
    return true;
  }

  return false;
};

// =====================================================
// 🔥 AUTO GROUP ACTS
// =====================================================

export const autoGroupActs = (
  acts: string[] = []
) => {

  const grouped:
  Record<string, string[]> = {};

  for (const act of acts) {

    const category =
      inferCategoryFromAct(act);

    if (!grouped[category]) {

      grouped[category] = [];
    }

    grouped[category].push(act);
  }

  return grouped;
};

// =====================================================
// 🔥 HELPERS
// =====================================================

export const isCriminalAct = (
  act: string
) =>
  inferCategoryFromAct(act)
  === "Criminal";

export const isCivilAct = (
  act: string
) =>
  inferCategoryFromAct(act)
  === "Civil";

export const isServiceLawAct = (
  act: string
) =>
  inferCategoryFromAct(act)
  === "Service Law";

export const isTaxCorporateAct = (
  act: string
) =>
  inferCategoryFromAct(act)
  === "Taxation & Corporate";

File: nlp_service/app/services/legal_ontology_service.py
import re

# =========================================================
# 🔥 CRIMINAL ACTS
# =========================================================

CRIMINAL_ACTS = [
    "Indian Penal Code",
    "Bharatiya Nyaya Sanhita",
    "Code Of Criminal Procedure",
    "Bharatiya Nagarik Suraksha Sanhita",
    "NDPS",
    "POCSO",
    "Prevention Of Corruption",
    "Prevention Of Money Laundering",
    "National Investigation Agency",
    "Food Adulteration",
    "Evidence Act",
    "Juvenile Justice",
    "UAPA",
    "TADA",
    "MCOCA",
]

# =========================================================
# 🔥 SERVICE LAW
# =========================================================

SERVICE_KEYWORDS = [
    "departmental proceeding",
    "dismissal",
    "reinstatement",
    "suspension",
    "promotion",
    "service matter",
    "pension",
    "disciplinary authority",
    "termination",
    "compulsory retirement",
]

# =========================================================
# 🔥 TAXATION / CORPORATE
# =========================================================

TAX_ACTS = [
    "Income Tax",
    "GST",
    "Goods And Services Tax",
    "Customs",
    "Central Excise",
    "Companies Act",
    "SEBI",
    "Insolvency",
    "SARFAESI",
    "Negotiable Instruments",
]

# =========================================================
# 🔥 CIVIL
# =========================================================

CIVIL_ACTS = [
    "Contract Act",
    "Transfer Of Property",
    "Specific Relief",
    "Arbitration",
    "Civil Procedure Code",
    "Constitution Of India",
    "Consumer Protection",
    "Motor Vehicles",
    "Wakf Act",
]

# =========================================================
# 🔥 ACT BASED CATEGORY
# =========================================================


def infer_category_from_act(acts):

    if not acts:
        return "Unknown"

    joined = " ".join([str(a).lower() for a in acts])

    for act in CRIMINAL_ACTS:

        if act.lower() in joined:
            return "Criminal"

    for act in TAX_ACTS:

        if act.lower() in joined:
            return "Taxation & Corporate"

    for act in CIVIL_ACTS:

        if act.lower() in joined:
            return "Civil"

    return "Unknown"


# =========================================================
# 🔥 SECTION BASED CATEGORY
# =========================================================


def infer_category_from_sections(sections):

    if not sections:
        return "Unknown"

    joined = str(sections).lower()

    criminal_sections = [
        "302",
        "307",
        "376",
        "420",
        "498a",
        "138",
        "125",
        "409",
        "467",
        "468",
        "471",
    ]

    for sec in criminal_sections:

        if sec in joined:
            return "Criminal"

    return "Unknown"


# =========================================================
# 🔥 CASE NUMBER CATEGORY
# =========================================================


def infer_category_from_case_number(case_number):

    if not case_number:
        return "Unknown"

    text = str(case_number).lower()

    if "criminal" in text:
        return "Criminal"

    if "civil" in text:
        return "Civil"

    if "service" in text:
        return "Service"

    if "tax" in text:
        return "Taxation & Corporate"

    if "writ petition" in text:
        return "Civil"

    if "special leave petition" in text:
        return "Civil"

    return "Unknown"


# =========================================================
# 🔥 ONTOLOGY ELIGIBILITY
# =========================================================


def is_ontology_eligible(data):

    case_number = str(data.get("caseNumber", "")).strip()

    citations = data.get("citations", [])

    doctrines = data.get("doctrines", [])

    acts = data.get("acts", [])

    points_of_law = data.get("points_of_law", [])

    final_holding = str(data.get("final_holding", "")).strip()

    dominant_issue = str(data.get("dominant_issue", "General")).strip()

    invalid_patterns = [
        r"^ITEM",
        r"^NO\.",
        r"SUPREME COURT OF INDIA NOTICE",
        r"UNKNOWN CASE",
    ]

    semantic_signals = any(
        [
            citations,
            doctrines,
            acts,
            points_of_law,
            final_holding not in ["", "Disposition Unknown"],
            dominant_issue not in ["", "General", "Unknown"],
        ]
    )

    if semantic_signals:
        return True

    for pattern in invalid_patterns:

        if re.search(pattern, case_number, re.I):
            return False

    if case_number:
        return True

    return False


# =========================================================
# 🔥 SEMANTIC CATEGORY CONFIDENCE ENGINE
# =========================================================


def build_category_confidence(acts=None, sections=None, case_number=""):

    acts = acts or []
    sections = sections or []

    category_scores = {
        "Criminal": 0,
        "Civil": 0,
        "Service": 0,
        "Taxation & Corporate": 0,
    }

    category_signals = []

    # =====================================================
    # 🔥 ACT SIGNALS
    # =====================================================

    acts_joined = " ".join([str(a).lower() for a in acts])

    for act in CRIMINAL_ACTS:

        if act.lower() in acts_joined:

            category_scores["Criminal"] += 35

            category_signals.append(f"Criminal Act: {act}")

    for act in TAX_ACTS:

        if act.lower() in acts_joined:

            category_scores["Taxation & Corporate"] += 35

            category_signals.append(f"Tax Act: {act}")

    for act in CIVIL_ACTS:

        if act.lower() in acts_joined:

            category_scores["Civil"] += 30

            category_signals.append(f"Civil Act: {act}")

    # =====================================================
    # 🔥 SECTION SIGNALS
    # =====================================================

    section_text = str(sections).lower()

    criminal_sections = ["302", "307", "376", "420", "498a", "467", "468", "471"]

    for sec in criminal_sections:

        if sec in section_text:

            category_scores["Criminal"] += 20

            category_signals.append(f"Criminal Section: {sec}")

    # =====================================================
    # 🔥 CASE NUMBER SIGNALS
    # =====================================================

    case_text = str(case_number).lower()

    if "criminal" in case_text:

        category_scores["Criminal"] += 40

        category_signals.append("Case Number: Criminal")

    if "civil" in case_text:

        category_scores["Civil"] += 40

        category_signals.append("Case Number: Civil")

    if "service" in case_text:

        category_scores["Service"] += 40

        category_signals.append("Case Number: Service")

    if "tax" in case_text:

        category_scores["Taxation & Corporate"] += 40

        category_signals.append("Case Number: Tax")

    # =====================================================
    # 🔥 FINAL CATEGORY
    # =====================================================

    top_category = max(category_scores, key=category_scores.get)

    top_confidence = category_scores[top_category]

    if top_confidence < 40:

        top_category = "Unknown"

    return {
        "category": top_category,
        "confidence": top_confidence,
        "scores": category_scores,
        "signals": category_signals,
    }
